kube-client/src/api/entry.rs

```diff
@@ -0,0 +1,161 @@
+            }),
+            None => Entry::Vacant(VacantEntry { api: self, name }),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Entry<'a, K> {
+    Occupied(OccupiedEntry<'a, K>),
+    Vacant(VacantEntry<'a, K>),
+}
+
+impl<'a, K> Entry<'a, K> {
+    pub fn get(&self) -> Option<&K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn get_mut(&mut self) -> Option<&mut K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get_mut()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn and_modify(self, f: impl FnOnce(&mut K)) -> Self {
+        match self {
+            Entry::Occupied(entry) => Entry::Occupied(entry.and_modify(f)),
+            entry @ Entry::Vacant(_) => entry,
+        }
+    }
+
+    pub fn or_insert(self, default: impl FnOnce() -> K) -> OccupiedEntry<'a, K>
+    where
+        K: Resource,
+    {
+        match self {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(entry) => entry.insert(default()),
+        }
+    }
+}
+
+pub struct OccupiedEntry<'a, K> {
+    api: &'a Api<K>,
+    dirtiness: Dirtiness,
+    object: K,
+}
+
+#[derive(Debug)]
+enum Dirtiness {
+    Clean,
+    Dirty,
+    New,
+}
+
+impl<'a, K> OccupiedEntry<'a, K> {
+    pub fn get(&self) -> &K {
+        &self.object
+    }
+
+    pub fn get_mut(&mut self) -> &mut K {
+        self.dirtiness = match self.dirtiness {
+            Dirtiness::Clean => Dirtiness::Dirty,
+            Dirtiness::Dirty => Dirtiness::Dirty,
+            Dirtiness::New => Dirtiness::New,
+        };
+        &mut self.object
+    }
+
+    pub fn and_modify(mut self, f: impl FnOnce(&mut K)) -> Self {
+        f(self.get_mut());
+        self
+    }
+
+    pub fn into_object(self) -> K {
+        self.object
+    }
+
+    pub async fn sync(&mut self) -> Result<()>
+    where
+        K: Resource + DeserializeOwned + Serialize + Clone + Debug,
+    {
+        self.object = match self.dirtiness {
+            Dirtiness::New => self.api.create(&PostParams::default(), &self.object).await?,
+            Dirtiness::Dirty => {
+                self.api
+                    .replace(
+                        self.object.meta().name.as_deref().unwrap(),
+                        &PostParams::default(),
+                        &self.object,
+                    )
+                    .await?
+            }
+            Dirtiness::Clean => self.api.get(self.object.meta().name.as_deref().unwrap()).await?,
+        };
+        self.dirtiness = Dirtiness::Clean;
+        Ok(())
+    }
+}
+
+pub struct VacantEntry<'a, K> {
+    api: &'a Api<K>,
+    name: &'a str,
+}
+
+impl<'a, K> VacantEntry<'a, K> {
+    pub fn insert(self, mut object: K) -> OccupiedEntry<'a, K>
+    where
+        K: Resource,
+    {
+        let meta = object.meta_mut();
+        meta.name.get_or_insert_with(|| self.name.to_string());
```

Comment:
this helpful insert needs to check for a `generateName` before deciding to potentially blanket override it

Reply:
Yeah, we should probably reject objects with `generateName` set entirely, since that doesn't really make sense for `get-and-create`. Maybe we should even reject objects where `obj.meta.name` != `self.name` as well (when set manually).

Reply:
Added warnings in 03a2967, is that fine or do we want to promote them to full-blown errors?

Reply:
i'm inclined to error on misuse of name/namespace/generate_name properties because the whole abstraction is geared toward giving a name then mutating it / inserting a new one in its place. It feels pretty strange to suddenly send in an entry with a different name, or in a namespace the api was not configured for.

but otoh, we would never catch all the cases where the user subverts any validation we put in place (via `get_mut` say).. so maybe the helpful warnings are enough? or maybe it's better to put the errors and the `.sync` stage and always store the `name` property on all `Entry` types

Reply:
Yeah.. this is is a tricky aspect to balance. Making `or_insert` fallible would be annoying ergonomically for what is a programming error in 99% of cases, but I'm also not sure I'm comfortable turning it into a `panic!`.

Moving the check into `sync` does make some sense...

Reply:
it's much better now :+1: